bitcoin/src/blockdata/transaction.rs

```diff
@@ -993,6 +993,83 @@ impl Transaction {
         // `Transaction` docs for full explanation).
         self.input.is_empty()
     }
+
+    /// Returns a reference to the input at `input_index` if it exists.
+    #[inline]
+    pub fn tx_in(&self, input_index: usize) -> Result<&TxIn, InputsIndexError> {
+        self.input
+            .get(input_index)
+            .ok_or(IndexOutOfBoundsError { index: input_index, length: self.input.len() }.into())
+    }
+
+    /// Returns a reference to the output at `output_index` if it exists.
+    #[inline]
+    pub fn tx_out(&self, output_index: usize) -> Result<&TxOut, OutputsIndexError> {
+        self.output
+            .get(output_index)
+            .ok_or(IndexOutOfBoundsError { index: output_index, length: self.output.len() }.into())
+    }
+}
+
+/// Error attempting to do an out of bounds access a the transaction inputs vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct InputsIndexError(pub IndexOutOfBoundsError);
+
+impl fmt::Display for InputsIndexError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write_err!(f, "invalid input index"; self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for InputsIndexError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
+}
+
+impl From<IndexOutOfBoundsError> for InputsIndexError {
+    fn from(e: IndexOutOfBoundsError) -> Self { Self(e) }
+}
+
+/// Error attempting to do an out of bounds access a the transaction outputs vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct OutputsIndexError(pub IndexOutOfBoundsError);
```

Comment:
There's a question whether this should be a named field since `#[non_exhaustive]` suggest we could add more fields but we'd have to keep it as tuple struct which wouldn't be great. We could remove `non_exhaustive` but there's a possible different choice: make `IndexOutOfBounds` error general-purpose, moving it from `transaction` module, so things like `txid` won't get added into that one but into this one. This is somewhat nice except for long error message.

Reply:
Oh I did not realize that we couldn't use `non_exhaustive` with tuple structs as you describe. There are a bunch of these in the code base now. Ugh `git grep -B 2 'pub struct .*Error'` shows we have non-uniform usage of `pub` on the inner filed as well.

Reply:
> we have non-uniform usage of `pub` on the inner filed as well.

That's fine, we don't have to commit to everything and should decide case-by-case.

Reply:
Ok, this PR is holding up other work. Can we just put it in like this? There are a bunch of these tuple struct error types in the codebase now with `non_exhaustive` on them that all have this problem. They all need fixing, adding this one is not making the code materially worse.

Reply:
Flagging this thread: Needs explicit ack from @Kixunil before merging.

Reply:
Yes, but needs a 1.0 issue.